content/en/blog/2024/java-metric-systems-compared/index.md

```diff
@@ -0,0 +1,491 @@
+
+**Figure 6:** Collect from exponential bucket histogram benchmark results.
+
+### Conclusions
+
+On the record side, micrometer and prometheus have an 11ns advantage for
+counters when attributes are known ahead of time, avoiding map lookups by using
+bound instruments to get direct references to aggregated state (not supported by
+OpenTelemetry). Despite this, OpenTelemetry has a 32ns advantage for explicit
+bucket histograms. This is likely due to micrometer and prometheus attempting to
+compute more summary values than the sum, min, max, and bucket counts of
+OpenTelemetry histograms. The advantage diminishes when attributes are not known
+ahead of time.
+
+When attributes are known ahead of time, none of the systems allocate memory
+when recording. This is great, and should be table stakes for any serious
+metrics system. When attributes are not known ahead of time (i.e. computed from
+application context), OpenTelemetry consistently allocates less memory than
+prometheus and micrometer. OpenTelemetry has clearly optimized for this scenario
+where micrometer and prometheus have a focus on attributes being known ahead of
+time and bound instruments. I argue that more often than not, attributes will
+not be known ahead of time, which marginalizes any advantage from micrometer or
+prometheus. Still, this is an area of potential improvement for OpenTelemetry.
+
+On the collect side, OpenTelemetry steals the show with extremely low memory
+allocation. When collecting without export, OpenTelemetry has anywhere from
+22%-99.7% fewer memory allocations than micrometer and prometheus. When
+collecting and exporting over OTLP, OpenTelemetry has anywhere from 85%-98.4%
+fewer memory allocations than micrometer and prometheus. Note that prometheus
+directly uses the OpenTelemetry OTLP exporter library, but without the
+optimization OpenTelemetry is able to achieve by being vertically integrated
+(i.e. exporter and core metrics system work together to achieve optimal
+results). Micrometer OTLP support translates the in-memory micrometer
+representation to generated Java classes before serializing, which is convenient
+but suboptimal from a performance standpoint.
+
+Overall, these are three serious metric systems. All perform admirably on the
+record side of things - a testament to the performance engineering that’s gone
+into all of them. After wrapping up a long series of optimizations,
+OpenTelemetry shines on the collect side of things. Its lower allocations will
+benefit every application, but are especially important to applications with
+high cardinality and with strict performance SLAs.
+
+If you’re reading this and considering Java metric systems, I hope you chose
+[opentelemetry-java][]. It’s a powerful and highly performant tool on its own,
+but comes with APIs for other key observability signals, a
+[rich instrumentation ecosystem](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/supported-libraries.md),
```

Comment:
I think we have a copy of that in docs as well. We also could point to the registry which might have instrumentation libraries outside the instrumentation repo: https://opentelemetry.io/ecosystem/registry/?language=java&component=instrumentation

Reply:
There's a section [here](https://opentelemetry.io/docs/languages/java/libraries/#available-instrumentation-libraries) but it just links to the page I linked to.

> https://opentelemetry.io/ecosystem/registry/?language=java&component=instrumentation

The registry could work, but for now the markdown page is better. It tells you whether library or autoinstrumentation is supported, and also which semantic conventions are produced. I would love it if the registry could do these things, but don't have time for it at the moment.

Reply:
Same here, it's a suggestion, so let's leave it as is